docs/container-affinity.md

```diff
@@ -0,0 +1,225 @@
+# Container Affinity and Anti-Affinity
+
+## Introduction
+
+Some policies allow the user to give hints about how particular containers
+should be *co-located* within a node. In particular these hints express whether
+containers should be located *'close'* to each other or *'far away'* from each
+other, in a hardware topology sense.
+
+Since these hints are interpreted always by a particular *policy implementation*,
+the exact definitions of 'close' and 'far' are also somewhat *policy-specific*.
+However as a general rule of thumb containers running
+
+  - on CPUs within the *same NUMA nodes* are considered *'close'* to each other,
+  - on CPUs within *different NUMA nodes* in the *same socket* are *'farther'*, and
+  - on CPUs within *different sockets* are *'far'* from each other
+
+These hints are expressed by `container affinity annotations` on the Pod.
+There are two types of affinities:
+
+  - `affinity` (or `positive affinty`): cause affected containers to *pull* each other closer
+  - `anti-affinity` (or `negative affinity`): cause affected containers to *push* each other further away
+
+Policies try to place a container
+  - close to those the container has affinity towards
+  - far from those the container has anti-affinity towards.
+
+## Affinity Annotation Syntax
+
+*Affinities* are defined as the `cri-resource-manager.intel.com/affinity` annotation.
+*Anti-affinities* are defined as the `cri-resource-manager.intel.com/anti-affinity`
+annotation. They are specified in the `metadata` section of the `Pod YAML`, under
+`annotations` as a dictionary, with each dictionary key being the name of the
+*container* within the Pod to which the annotation belongs to.
+
+```
+metadata:
+  anotations:
+    cri-resource-manager.intel.com/affinity: |
+      container1:
+        - scope:
+            key: [optional-key-domain/]key
+            operator: op
+            values:
+            - value1
+            ...
+            - valueN
+        - match:
+            key: [optional-key-domain/]key
+            operator: op
+            values:
+            - value1
+            ...
+            - valueN
+          weight: w
+```
+
+An anti-affinity is defined similarly but using `cri-resource-manager.intel.com/anti-affinity`
+as the annotation key.
+
+```
+metadata:
+  anotations:
+    cri-resource-manager.intel.com/anti-affinity: |
+      container1:
+        - scope:
+            key: [optional-key-domain/]key
+            operator: op
+            values:
+            - value1
+            ...
+            - valueN
+        - match:
+            key: [optional-key-domain/]key
+            operator: op
+            values:
+            - value1
+            ...
+            - valueN
+          weight: w
+```
+
+## Affinity Semantics
+
+An affinity consists of three parts:
+
+  - `scope`: defines which containers this affinity is evaluated against
+  - `expression`: defines for which containers (within the scope) the affinity applies to
```

Comment:
Just for coherency: should we use term "`match` expression (just *expression* later in this document)" or something similar here as the key used in the annotation is *match*?

Reply:
I changed both 'scope' and 'expression' to 'scope expression' and 'match expression', since they are both expressions (as a bit later stated in the document).